### PR TITLE
Create a basic histogram and timer structs for collecting statistics

### PR DESCRIPTION
**Pull request by andrisaar**
Although we can't measure actual seconds, counting clock cycles via `RDTSC` should be good enough to get an idea where time is spent.

### CODE REVIEW

File: oak_core/src/timer.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        let stop = rdtsc();
+        stop - self.start
+    }
+}
+
+/// Basic histogram to compute percentiles over some measurement.
+/// N is the number of elements kept in memory; it must be larger than 0.
+/// The array is allocated statically; Histogram does not use heap allocation.
+#[derive(Debug, PartialEq)]
+pub struct Histogram<const N: usize> {
```
Comment (conradgrobler) [inline oak_core/src/timer.rs:75]: nit: I don't think that `Histogram` is an accurate name here. Based on the name I initially assumed this would count the number of values falling in different ranges, so it took me quite some time to figure out what the code was doing. Perhaps something like `SampleStore` would be better.

File: oak_core/src/timer.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    pub fn percentile(&self, n: u8) -> Option<u64> {
+        if n > 100 || self.samples == 0 {
+            return None;
+        }
+        let mut array = self.data;
+        // If we've had more samples, cap the size of the slice at N.
+        let size = core::cmp::min(N, self.samples);
+        let slice = &mut array[..size];
+        slice.sort_unstable();
+        let index = ((size - 1) * n as usize) / 100;
```
Comment (conradgrobler) [inline oak_core/src/timer.rs:127]: I am a bit unsure about this. It feels like this might introduce some bias, but I don't immdiately have a better suggestion. Perhaps you could add a test or two to cover edge cases, specificallly with large sizes.

**Review by conradgrobler — APPROVED** [review]

File: oak_core/src/timer.rs (hunk 2 above)
Comment (tiziano88) [inline oak_core/src/timer.rs:127]: Perhaps use checked operations throughout.

File: oak_core/src/timer.rs (hunk 1 above)
Comment (tiziano88) [inline oak_core/src/timer.rs:75]: This would probably belong to a separate file, conceptually it's not specific to timer (and we may well reuse it for other stuff in the future).

**Review by tiziano88 — APPROVED** [review]

File: oak_core/src/timer.rs (hunk 2 above)
Comment (andrisaar) [reply oak_core/src/timer.rs:127]: I've changed it to use floats (so that we could express 99.9%, for example), so this is now moot.

Comment (andrisaar) [reply oak_core/src/timer.rs:127]: I added a test with a store of 1000, and it seems to work, at least when filled with predictable data.
